bridge/github/import_mediator.go

```diff
@@ -0,0 +1,394 @@
+	frame, _ := frames.Next()
+	fmt.Printf("%s:%d %s\n", frame.File, frame.Line, frame.Function)
+}
+
+const (
+	NUM_ISSUES         = 50
+	NUM_ISSUE_EDITS    = 99
+	NUM_TIMELINE_ITEMS = 99
+	NUM_COMMENT_EDITS  = 99
+
+	CHAN_CAPACITY = 128
+)
+
+// TODO: remove all debug output and trace() in all files. Use ag
+
+type importMediator struct {
+	// Github graphql client
+	gc      *githubv4.Client
+	owner   string
+	project string
+	// The iterator will only query issues updated or created after the date given in
+	// the variable since.
+	since time.Time
+
+	issues           chan issue
+	issueEditsMut    sync.Mutex
+	timelineItemsMut sync.Mutex
+	commentEditsMut  sync.Mutex
+	issueEdits       map[githubv4.ID]chan userContentEdit
+	timelineItems    map[githubv4.ID]chan timelineItem
+	commentEdits     map[githubv4.ID]chan userContentEdit
+
+	// Sticky error
+	err error
+}
+
+func NewImportMediator(ctx context.Context, client *githubv4.Client, owner, project string, since time.Time) *importMediator {
+	mm := importMediator{
+		gc:               client,
+		owner:            owner,
+		project:          project,
+		since:            since,
+		issues:           make(chan issue, CHAN_CAPACITY),
+		issueEditsMut:    sync.Mutex{},
+		timelineItemsMut: sync.Mutex{},
+		commentEditsMut:  sync.Mutex{},
+		issueEdits:       make(map[githubv4.ID]chan userContentEdit),
+		timelineItems:    make(map[githubv4.ID]chan timelineItem),
+		commentEdits:     make(map[githubv4.ID]chan userContentEdit),
+		err:              nil,
+	}
+	go func() {
+		defer close(mm.issues)
+		mm.fillChannels(ctx)
+	}()
+	return &mm
+}
+
+func (mm *importMediator) Issues() <-chan issue {
+	return mm.issues
+}
+
+func (mm *importMediator) IssueEdits(issue *issue) <-chan userContentEdit {
+	mm.issueEditsMut.Lock()
+	channel := mm.issueEdits[issue.Id]
+	mm.issueEditsMut.Unlock()
+	return channel
+}
+
+func (mm *importMediator) TimelineItems(issue *issue) <-chan timelineItem {
+	mm.timelineItemsMut.Lock()
+	channel := mm.timelineItems[issue.Id]
+	mm.timelineItemsMut.Unlock()
+	return channel
+}
+
+func (mm *importMediator) CommentEdits(comment *issueComment) <-chan userContentEdit {
+	mm.commentEditsMut.Lock()
+	channel := mm.commentEdits[comment.Id]
+	mm.commentEditsMut.Unlock()
+	return channel
+}
+
+func (mm *importMediator) Error() error {
+	return mm.err
+}
+
+func (mm *importMediator) User(ctx context.Context, loginName string) (*user, error) {
+	query := userQuery{}
+	vars := varmap{"login": githubv4.String(loginName)}
+	c, cancel := context.WithTimeout(ctx, defaultTimeout)
+	defer cancel()
+	if err := mm.mQuery(c, &query, vars); err != nil {
+		return nil, err
+	}
+	return &query.User, nil
+}
+
+func (mm *importMediator) fillChannels(ctx context.Context) {
+	issueCursor := githubv4.String("")
+	for {
+		issues, hasIssues := mm.queryIssue(ctx, issueCursor)
+		if !hasIssues {
+			break
+		}
+		issueCursor = issues.PageInfo.EndCursor
+		for _, issueNode := range issues.Nodes {
```

Comment:
I might be wrong but when you have a loop pushing to a channel like that, you might need to wrap it a `select` that also monitor the context, so that the loop can exit properly when needed.

Something like that:
```go
	for _, foo := range bar {
		// do something with foo
		foo2 := foo + "hello"
		
		select {
		case <-ctx.Done():
			return
		case out<-foo2:
		}
	}
	```

Reply:
This might explain why the import doesn't stop properly like in https://github.com/MichaelMure/git-bug/pull/585#issuecomment-789200158

Reply:
Yes, you are right, I included it in one of my commits.